Updated release notes

imobachgs opened the pull request:
This PR allows to fetch release notes using the old-mechanism (`relnotes_url` property). You can find a description of how it works now in the [Y2Packager::ReleaseNotesReader class](https://github.com/yast/yast-packager/blob/d043a487665811706edcf4c92039503a0b5cb627/src/lib/y2packager/release_notes_reader.rb#L22).

The code to download release notes in the old-way was copied and adapted from the [InstDownloadReleaseNotes](https://github.com/yast/yast-installation/blob/62596684d6de242667a0957765c85712874e77ef/src/lib/installation/clients/inst_download_release_notes.rb) that will be replaced pretty soon for a newer version.

Unit testing coverage is pretty high for the new code, but I would like to do some real tests before considering it done.

mvidner commented on src/lib/y2packager/release_notes_reader.rb:
```diff
@@ -11,170 +11,133 @@
+        if registered?
+          [
+            ReleaseNotesRpmReader.new(product),
+            ReleaseNotesUrlReader.new(product)
```
Well, this suggests that these 2 classes should have a superclass that documents their interface
(and the #latest_version type of [String,:latest] would clarify my earlier question).

But an obvious name ReleaseNotesReader is already taken for something else.

imobachgs replied:
Yes, we could use `ReleaseNotesFetcher` (with `ReleaseNotesRpmFetcher` and `ReleaseNotesUrlFetcher`) or `ReleaseNotesStrategy` or something like that. I will try to come up with some name which makes sense. I am also open to suggestions :)

mvidner commented on src/lib/y2packager/release_notes_reader.rb:
```diff
@@ -11,170 +11,133 @@
+    # @param reader        [ReleaseNotesRpmReader,ReleaseNotesUrlReader] Release notes reader
+    # @param user_lang     [String] User preferred language (falling back to fallback_lang)
+    # @param format        [Symbol] Release notes format (:txt or :rtf)
+    # @param fallback_lang [String] Release notes fallback language
```
I see this triple of user_lang, format, fallback_lang in so many places that I think it might be better to encapsulate it in a ContentPreference object?

imobachgs replied:
Yes, it makes sense.

mvidner commented on src/lib/y2packager/release_notes_url_reader.rb:
```diff
@@ -0,0 +1,371 @@
+    }.freeze
+
+    # Product to get release notes for
+    attr_reader :product
```
Well, that's rather my obsession with type checking. The link could be done easier with `{Product} to get ...`

imobachgs commented on src/lib/y2packager/release_notes_reader.rb:
```diff
@@ -11,170 +11,133 @@
+  # ### Cached release notes
+  #
+  # Release notes are stored using an instance of `Y2Packager::ReleaseNotesStore`.
+  # When trying to read a product license for second time, this class will try to fetch
```
Just a mental typo. I was working in the "licenses" stuff for two sprints in a row :)

imobachgs commented on src/lib/y2packager/release_notes_rpm_reader.rb:
```diff
@@ -0,0 +1,178 @@
+      Y2Packager::Package
+        .find(name)
+        .select { |i| AVAILABLE_STATUSES.include?(i.status) }
+        .sort_by { |i| Gem::Version.new(i.version) }
```
The I would need to find a better way (and try to make it reusable). We already [use Gem::Version](https://github.com/yast/yast-migration/blob/7493de13d94d48b8af5f00dd8465edc1c4beeb0f/src/lib/migration/repository_checker.rb#L86) but for products (and version numbers are different there).

imobachgs commented:
@mvidner Thanks for the review! I've updated the PR according to your comments. Summarizing:

* Some documentation fixes
* Moved/Renamed `Y2Packager::ReleaseNotesRpmReader` and `Y2Packager::ReleaseNotesUrlReader` to `Y2Packager::ReleaseNotesFetchers::Rpm` and `Y2Packager::ReleaseNotesFetchers::Url`.
* Defined a new base class for fetchers (`Y2Packager::ReleaseNotesFetchers::Base`).
* Added a new `Y2Packager::ReleaseNotesContentPrefs` to store release notes content preferences.
* Use a new `Yast::Pkg.CompareVersions` to compare package versions.

lslezak commented on src/lib/y2packager/release_notes_fetchers/url.rb:
```diff
@@ -0,0 +1,360 @@
+            log.info("Release notes index empty, not filtering further downloads")
+            nil
+          else
+            rn_filter = index_file.split("\n")
```
NP: `rn_filter` sounds like a Proc for filtering release notes, I'd use `rn_index` or something like that here.

lslezak commented on src/lib/y2packager/release_notes_fetchers/url.rb:
```diff
@@ -0,0 +1,360 @@
+          max_time
+        )
+        ret = Yast::SCR.Execute(Yast::Path.new(".target.bash"), cmd)
+        log.info("#{cmd} returned #{ret}")
```
NP: I'd remove `curl_proxy_args` from the logged string (`cmd.sub(curl_proxy_args, "")`), logging proxy passwords should be avoided.  (The `y2log` might be attached to bugzilla and the private data could leak.)

lslezak replied:
Or replace with "@PROXYPASSWORD@" or something like that.

lslezak commented on src/lib/y2packager/release_notes_reader.rb:
```diff
@@ -11,170 +11,131 @@
+  # We can distinguish two different case:
+  #
+  # * When the system *is registered*: release notes will be obtained from RPM packages.
+  #   If release notes are not found there, it will fall back to the relnotes_url property.
```
NP: `"relnotes_url" product property`, to make clear what it is about...

lslezak commented on src/lib/y2packager/release_notes_reader.rb:
```diff
@@ -11,170 +11,131 @@
+  # or the stored version is outdated (maybe a new package is now available), it will
+  # try to get that version.
+  #
+  # Take into account that, when using the relnotes_url property, a URL that already
```
Typo: **an** URL

lslezak commented on test/lib/release_notes_fetchers/url_test.rb:
```diff
@@ -0,0 +1,275 @@
+
+        it "uses an unauthenticated proxy" do
+          expect(Yast::SCR).to receive(:Execute) do |_path, cmd|
+            expect(cmd).to include("--proxy http://proxy.example.com")
```
The `--proxy` option is used also with an authenticated proxy (see the test below). The correct test would be the `--proxy` option is present and the `--proxy-user` option is **not** present.

lslezak commented on test/lib/release_notes_reader_test.rb:
```diff
@@ -2,160 +2,165 @@
-        end
+      it "retrieves release notes from external sources" do
+        expect(url_reader).to receive(:release_notes).with(
+          Y2Packager::ReleaseNotesContentPrefs.new("cz_CZ", "en", :txt)
```
JFYI: the language code is `cs`, the country code is `CZ`. Yes a bit confusing but the ISO language code uses that given language for the code, like for German it's `de`.... so the correct string here is `cs_CZ`.

lslezak commented on test/lib/release_notes_reader_test.rb:
```diff
@@ -2,160 +2,165 @@
+          expect(url_reader).to receive(:release_notes).with(
+            Y2Packager::ReleaseNotesContentPrefs.new("cz_CZ", "en", :html)
+          )
+          rn = reader.release_notes(user_lang: "cz_CZ", format: :html)
```
...and the same here...

lslezak commented on src/lib/y2packager/release_notes_fetchers/url.rb:
```diff
@@ -346,7 +346,8 @@ def curl_download(url, filename, max_time: 300)
         )
         ret = Yast::SCR.Execute(Yast::Path.new(".target.bash"), cmd)
-        log.info("#{cmd} returned #{ret}")
+        filtered_cmd = cmd.sub(/--proxy-user '[^']*'/, "--proxy-user @PROXYPASSWORD@")
```
You can use `curl_proxy_args` string directly (it is used few lines above), simply `cmd.sub(curl_proxy_args, "--proxy-user @PROXYPASSWORD@")`

imobachgs commented:
Thanks a lot!
